File: src/commands/payload.cpp
#include "payload/payload.h"
#include "obc.h"
#include "payload/devices.h"

using std::uint16_t;

using namespace drivers::payload;

typedef void (*VoidFuncPtr)(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[]);

static void ShowHelp()
{
    Main.terminal.Printf("payload temps|photo|suns|house|who|radfet on|radfet read|radfet off\n");
}

static void RadFET(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[])
{
    if (argc < 1)
    {
        ShowHelp();
        return;
    }

    auto option = argv[0];

    PayloadTelemetry::Radfet result;
    OSResult status = OSResult::InvalidArgument;

    if (strcmp(option, "on") == 0)
    {
        Main.terminal.Printf("Starting RadFet.\n");
        status = driver.RadFETOn(result);
    }
    else if (strcmp(option, "read") == 0)
    {
        Main.terminal.Printf("Starting RadFet Measurements.\n");
        status = driver.MeasureRadFET(result);
    }
    else if (strcmp(option, "off") == 0)
    {
        Main.terminal.Printf("Stopping RadFet.\n");
        status = driver.RadFETOff(result);
    }

    if (status != OSResult::Success)
    {
        Main.terminal.Printf("RadFet Test failed!\n");
        return;
    }
    else
    {
        Main.terminal.Printf("Status: %d\n", result.status);
        Main.terminal.Printf("Temperature: %ld\n", result.temperature);
        Main.terminal.Printf("Vth0: %ld\n", result.vth[0]);
        Main.terminal.Printf("Vth1: %ld\n", result.vth[1]);
        Main.terminal.Printf("Vth2: %ld\n", result.vth[2]);
        Main.terminal.Printf("Success!");
    }
}

static void Temperatures(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[])
{
    UNREFERENCED_PARAMETER(argc);
    UNREFERENCED_PARAMETER(argv);

    PayloadTelemetry::Temperatures result;

    Main.terminal.Printf("Starting Temperatures Test!\n");
    auto status = driver.MeasureTemperatures(result);

    if (status != OSResult::Success)
    {
        Main.terminal.Printf("Temperatures Test failed!\n");
        return;
    }
    else
    {
        Main.terminal.Printf("Supply: %d\n", result.supply);
        Main.terminal.Printf("Xp: %d\n", result.Xp);
        Main.terminal.Printf("Xn: %d\n", result.Xn);
        Main.terminal.Printf("Yp: %d\n", result.Yp);
        Main.terminal.Printf("Yn: %d\n", result.Yn);
        Main.terminal.Printf("SADS: %d\n", result.sads);
        Main.terminal.Printf("Sail: %d\n", result.sail);
        Main.terminal.Printf("CAM Nadir: %d\n", result.cam_nadir);
        Main.terminal.Printf("CAM Wing: %d\n", result.cam_wing);
        Main.terminal.Printf("Success!");
    }
}

static void Photodiodes(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[])
{
    UNREFERENCED_PARAMETER(argc);
    UNREFERENCED_PARAMETER(argv);

    PayloadTelemetry::Photodiodes result;

    Main.terminal.Printf("Starting Photodiode Test!\n");
    auto status = driver.MeasurePhotodiodes(result);

    if (status != OSResult::Success)
    {
        Main.terminal.Printf("Photodiode Test failed!\n");
        return;
    }
    else
    {
        Main.terminal.Printf("Xp: %d\n", result.Xp);
        Main.terminal.Printf("Xn: %d\n", result.Xn);
        Main.terminal.Printf("Yp: %d\n", result.Yp);
        Main.terminal.Printf("Yn: %d\n", result.Yn);
        Main.terminal.Printf("Success!");
    }
}

static void SunS(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[])
{
    UNREFERENCED_PARAMETER(argc);
    UNREFERENCED_PARAMETER(argv);

    PayloadTelemetry::SunsRef result;

    Main.terminal.Printf("Starting SunS Test!\n");
    auto status = driver.MeasureSunSRef(result);

    if (status != OSResult::Success)
    {
        Main.terminal.Printf("Suns Test failed!\n");
        return;
    }
    else
    {
        Main.terminal.Printf("V1: %d\n", result.voltages[0]);
        Main.terminal.Printf("V2: %d\n", result.voltages[1]);
        Main.terminal.Printf("V3: %d\n", result.voltages[2]);
        Main.terminal.Printf("V4: %d\n", result.voltages[3]);
        Main.terminal.Printf("V5: %d\n", result.voltages[4]);
        Main.terminal.Printf("Success!");
    }
}

static void Housekeeping(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[])
{
    UNREFERENCED_PARAMETER(argc);
    UNREFERENCED_PARAMETER(argv);

    PayloadTelemetry::Housekeeping result;

    Main.terminal.Printf("Starting Housekeeping Test!\n");
    auto status = driver.MeasureHousekeeping(result);

    if (status != OSResult::Success)
    {
        Main.terminal.Printf("Housekeeping Test failed!\n");
        return;
    }
    else
    {
        Main.terminal.Printf("INT 3V3D: %d\n", result.int_3v3d);
        Main.terminal.Printf("OBC 3V3D: %d\n", result.obc_3v3d);
        Main.terminal.Printf("Success!");
    }
}

static void Whoami(IPayloadDeviceDriver& driver, uint16_t argc, char* argv[])
{
    UNREFERENCED_PARAMETER(argc);
    UNREFERENCED_PARAMETER(argv);

    PayloadTelemetry::Status result;

    Main.terminal.Printf("Starting Whoami Test!\n");
    auto status = driver.GetWhoami(result);

    if (status != OSResult::Success)
    {
        Main.terminal.Printf("WhoAmI Test failed!\n");
        return;
    }
    else
    {
        Main.terminal.Printf("Who Am I: %d (0x%x)\n", result.who_am_i, result.who_am_i);
        Main.terminal.Printf("Success!");
    }
}

static VoidFuncPtr GetDriverCommand(char* name)
{
    if (strcmp(name, "radfet") == 0)
    {
        return RadFET;
    }
    else if (strcmp(name, "temps") == 0)
    {
        return Temperatures;
    }
    else if (strcmp(name, "photo") == 0)
    {
        return Photodiodes;
    }
    else if (strcmp(name, "suns") == 0)
    {
        return SunS;
    }
    else if (strcmp(name, "house") == 0)
    {
        return Housekeeping;
    }
    else if (strcmp(name, "who") == 0)
    {
        return Whoami;
    }
    else
        return nullptr;
}

void PayloadDriver(uint16_t argc, char* argv[])
{
    if (argc == 0)
    {
        ShowHelp();
        return;
    }
    auto cmd = GetDriverCommand(argv[0]);
    if (cmd == nullptr)
    {
        ShowHelp();
        return;
    }

    PayloadDeviceDriver driver(Main.Hardware.PayloadDriver);

    cmd(driver, argc - 1, &argv[1]);
}

File: integration_tests/devices/payload.py
import struct
import i2cMock
from i2cMock import I2CDevice
from threading import Timer


class Payload(I2CDevice):
    def __init__(self, gpio_driver, pin):
        super(Payload, self).__init__(0b0110000, "Payload")

        self.interrupt_pin = pin
        self.gpioDriver = gpio_driver
        self.Status = self.StatusTelemetry()
        self.SunS_Ref = self.SunSTelemetry()
        self.Temperatures = self.TemperatureTelemetry()
        self.Photodiodes = self.PhotodiodeTelemetry()
        self.Housekeeping = self.HousekeepingTelemetry()
        self.RadFET = self.RadFETTelemetry()

        # processing hooks
        self.timeout_callback = None
        self.external_measure_suns_ref_callback = None
        self.external_measure_temperatures_callback = None
        self.external_measure_photodiodes_callback = None
        self.external_measure_housekeeping_callback = None
        self.external_initialize_radfet_callback = None
        self.external_measure_radfet_callback = None
        self.external_finish_radfet_callback = None

    # Interrupt pin handling

    def gpio_interrupt_low(self):
        self.gpioDriver.gpio_low(self.interrupt_pin)

    def gpio_interrupt_high(self):
        self.gpioDriver.gpio_high(self.interrupt_pin)

    # Commands

    @i2cMock.command([0x80])
    def measure_suns(self):
        if self.external_measure_suns_ref_callback:
            self.external_measure_suns_ref_callback()
        else:
            self.mock_measure_suns_ref()
            self.mock_processing_start()

    @i2cMock.command([0x81])
    def measure_temperatures(self):
        if self.external_measure_temperatures_callback:
            self.external_measure_temperatures_callback()
        else:
            self.mock_measure_temperatures()
            self.mock_processing_start()

    @i2cMock.command([0x82])
    def measure_photodiodes(self):
        if self.external_measure_photodiodes_callback:
            self.external_measure_photodiodes_callback()
        else:
            self.mock_measure_photodiodes()
            self.mock_processing_start()

    @i2cMock.command([0x83])
    def measure_housekeeping(self):
        if self.external_measure_housekeeping_callback:
            self.external_measure_housekeeping_callback()
        else:
            self.mock_measure_housekeeping()
            self.mock_processing_start()

    @i2cMock.command([0x84])
    def radfet_on(self):
        if self.external_initialize_radfet_callback:
            self.external_initialize_radfet_callback()
        else:
            self.mock_initialize_radfet()
            self.mock_processing_start()

    @i2cMock.command([0x85])
    def measure_radfet(self):
        if self.external_measure_radfet_callback:
            self.external_measure_radfet_callback()
        else:
            self.mock_measure_radfet()
            self.mock_processing_start()

    @i2cMock.command([0x86])
    def radfet_off(self):
        if self.external_finish_radfet_callback:
            self.external_finish_radfet_callback()
        else:
            self.mock_finish_radfet()
            self.mock_processing_start()

    # Data Reads

    @i2cMock.command([0x00])
    def read_whoami(self):
        return self.Status.get()

    @i2cMock.command([0x01])
    def read_suns_ref(self):
        return self.SunS_Ref.get()

    @i2cMock.command([11])
    def read_temperatures(self):
        return self.Temperatures.get()

    @i2cMock.command([29])
    def read_photodiodes(self):
        return self.Photodiodes.get()

    @i2cMock.command([37])
    def read_housekeeping(self):
        return self.Housekeeping.get()

    @i2cMock.command([41])
    def read_radfet(self):
        return self.RadFET.get()

    # Value setters

    def set_suns_ref(self, new_value):
        self.SunS_Ref.set(new_value)

    def set_temperatures(self, new_value):
        self.Temperatures.set(new_value)

    def set_photodiodes(self, new_value):
        self.Photodiodes.set(new_value)

    def set_housekeeping(self, new_value):
        self.Housekeeping.set(new_value)

    def set_radfet(self, new_value):
        self.RadFET.set(new_value)

    def set_whoami(self, new_value):
        self.Status.set(new_value)

    def set_timeout_callback(self, new_timeout_callback=None):
        self.timeout_callback = new_timeout_callback

    def set_external_measure_suns_ref_callback(self, new_value=None):
        self.external_measure_suns_ref_callback = new_value

    def set_external_measure_temperatures_callback(self, new_value=None):
        self.external_measure_temperatures_callback = new_value

    def set_external_measure_photodiodes_callback(self, new_value=None):
        self.external_measure_photodiodes_callback = new_value

    def set_external_measure_housekeeping_callback(self, new_value=None):
        self.external_measure_housekeeping_callback = new_value

    def set_external_initialize_radfet_callback(self, new_value=None):
        self.external_initialize_radfet_callback = new_value

    def set_external_measure_radfet_callback(self, new_value=None):
        self.external_measure_radfet_callback = new_value

    def set_external_finish_radfet_callback(self, new_value=None):
        self.external_finish_radfet_callback = new_value

    # Measurements mock

    def mock_processing_start(self):
        self.gpio_interrupt_high()
        self.log.info("Start PLD measurement")

        timeout = 0
        if self.timeout_callback is not None:
            timeout = self.timeout_callback()
        t = Timer(timeout, self.mock_processing_stop)
        t.start()

    def mock_processing_stop(self):
        self.log.info("Finished PLD measurement")
        self.gpio_interrupt_low()

    def mock_measure_suns_ref(self):
        if self.SunS_Ref.is_empty():
            self.set_suns_ref({'V1': 101, 'V2': 102, 'V3': 103, 'V4': 104, 'V5': 105})

    def mock_measure_temperatures(self):
        if self.Temperatures.is_empty():
            self.set_temperatures({'Supply': 201, 'Xp': 202, 'Xn': 203, 'Yp': 204, 'Yn': 205, 'SADS': 206,
                                   'Sail': 207, 'CAM Nadir': 208, 'CAM Wing': 209})

    def mock_measure_photodiodes(self):
        if self.Photodiodes.is_empty():
            self.set_photodiodes({'Xp': 301, 'Xn': 302, 'Yp': 303, 'Yn': 304})

    def mock_measure_housekeeping(self):
        if self.Housekeeping.is_empty():
            self.set_housekeeping({'INT 3V3D': 401, 'OBC 3V3D': 402})

    def mock_initialize_radfet(self):
        self.set_radfet({'Status': 50, 'Temperature': 500001, 'Vth0': 500002,
                         'Vth1': 500003, 'Vth2': 500004})

    def mock_measure_radfet(self):
        if not self.RadFET.is_empty():
            self.RadFET.Status = self.RadFET.Status + 1
            self.RadFET.Temperature = self.RadFET.Temperature + 100
            self.RadFET.Vth0 = self.RadFET.Vth0 + 100
            self.RadFET.Vth1 = self.RadFET.Vth1 + 100
            self.RadFET.Vth2 = self.RadFET.Vth2 + 100
        else:
            # mock read of unintialized radfet
            self.set_radfet({'Status': 99, 'Temperature': 0x00FFFFFF, 'Vth0': 0x00FFFFFF,
                             'Vth1': 0x00FFFFFF, 'Vth2': 0x00FFFFFF})

    def mock_finish_radfet(self):
        if not self.RadFET.is_empty():
            self.set_radfet({'Status': 59, 'Temperature': 590001, 'Vth0': 590002,
                             'Vth1': 590003, 'Vth2': 590004})
        else:
            # mock read of unintialized radfet
            self.set_radfet({'Status': 111, 'Temperature': 0x00FFFFFF, 'Vth0': 0x00FFFFFF,
                             'Vth1': 0x00FFFFFF, 'Vth2': 0x00FFFFFF})

    # Telemetry classes

    class StatusTelemetry:
        def __init__(self):
            self.WhoAmI = 0x53

        def set(self, telemetry_dict):
            self.WhoAmI = telemetry_dict.get('Who Am I')

        def get(self):
            return list(struct.pack('<B', self.WhoAmI))

    class SunSTelemetry:
        def __init__(self):
            self.V1 = 0xFFFF
            self.V2 = 0xFFFF
            self.V3 = 0xFFFF
            self.V4 = 0xFFFF
            self.V5 = 0xFFFF

        def set(self, telemetry_dict):
            self.V1 = telemetry_dict.get('V1')
            self.V2 = telemetry_dict.get('V2')
            self.V3 = telemetry_dict.get('V3')
            self.V4 = telemetry_dict.get('V4')
            self.V5 = telemetry_dict.get('V5')

        def get(self):
            return list(struct.pack('<HHHHH',
                                    self.V1,
                                    self.V2,
                                    self.V3,
                                    self.V4,
                                    self.V5))

        def is_empty(self):
            return self.V1 == 0xFFFF and self.V2 == 0xFFFF and self.V3 == 0xFFFF \
                   and self.V4 == 0xFFFF and self.V5 == 0xFFFF

    class TemperatureTelemetry:
        def __init__(self):
            self.Supply = 0xFFFF
            self.Xp = 0xFFFF
            self.Xn = 0xFFFF
            self.Yp = 0xFFFF
            self.Yn = 0xFFFF
            self.SADS = 0xFFFF
            self.Sail = 0xFFFF
            self.CAMNadir = 0xFFFF
            self.CAMWing = 0xFFFF

        def set(self, telemetry_dict):
            self.Supply = telemetry_dict.get('Supply')
            self.Xp = telemetry_dict.get('Xp')
            self.Xn = telemetry_dict.get('Xn')
            self.Yp = telemetry_dict.get('Yp')
            self.Yn = telemetry_dict.get('Yn')
            self.SADS = telemetry_dict.get('SADS')
            self.Sail = telemetry_dict.get('Sail')
            self.CAMNadir = telemetry_dict.get('CAM Nadir')
            self.CAMWing = telemetry_dict.get('CAM Wing')

        def get(self):
            return list(struct.pack('<HHHHHHHHH',
                                    self.Supply,
                                    self.Xp,
                                    self.Xn,
                                    self.Yp,
                                    self.Yn,
                                    self.SADS,
                                    self.Sail,
                                    self.CAMNadir,
                                    self.CAMWing))

        def is_empty(self):
            return self.Supply == 0xFFFF and self.Xp == 0xFFFF and self.Xn == 0xFFFF and self.Yp == 0xFFFF \
                   and self.Yn == 0xFFFF and self.SADS == 0xFFFF and self.Sail == 0xFFFF and self.CAMNadir == 0xFFFF \
                   and self.CAMWing == 0xFFFF

    class PhotodiodeTelemetry:
        def __init__(self):
            self.Xp = 0xFFFF
            self.Xn = 0xFFFF
            self.Yp = 0xFFFF
            self.Yn = 0xFFFF

        def set(self, telemetry_dict):
            self.Xp = telemetry_dict.get('Xp')
            self.Xn = telemetry_dict.get('Xn')
            self.Yp = telemetry_dict.get('Yp')
            self.Yn = telemetry_dict.get('Yn')

        def get(self):
            return list(struct.pack('<HHHH',
                                    self.Xp,
                                    self.Xn,
                                    self.Yp,
                                    self.Yn))

        def is_empty(self):
            return self.Xp == 0xFFFF and self.Xn == 0xFFFF and self.Yp == 0xFFFF and self.Yn == 0xFFFF

    class HousekeepingTelemetry:
        def __init__(self):
            self.INT_3V3D = 0xFFFF
            self.OBC_3V3D = 0xFFFF

        def set(self, telemetry_dict):
            self.INT_3V3D = telemetry_dict.get('INT 3V3D')
            self.OBC_3V3D = telemetry_dict.get('OBC 3V3D')

        def get(self):
            return list(struct.pack('<HH',
                        self.INT_3V3D,
                        self.OBC_3V3D))

        def is_empty(self):
            return self.INT_3V3D == 0xFFFF and self.OBC_3V3D == 0xFFFF

    class RadFETTelemetry:
        def __init__(self):
            self.Status = 0xFF
            self.Temperature = 0xFFFFFFFF
            self.Vth0 = 0xFFFFFFFF
            self.Vth1 = 0xFFFFFFFF
            self.Vth2 = 0xFFFFFFFF

        def set(self, telemetry_dict):
            self.Status = telemetry_dict.get('Status')
            self.Temperature = telemetry_dict.get('Temperature')
            self.Vth0 = telemetry_dict.get('Vth0')
            self.Vth1 = telemetry_dict.get('Vth1')
            self.Vth2 = telemetry_dict.get('Vth2')

        def get(self):
            return list(struct.pack('<BLLLL',
                                    self.Status,
                                    self.Temperature,
                                    self.Vth0,
                                    self.Vth1,
                                    self.Vth2))

        def is_empty(self):
            return self.Status == 0xFF and self.Temperature == 0xFFFFFFFF and self.Vth0 == 0xFFFFFFFF \
                   and self.Vth1 == 0xFFFFFFFF and self.Vth2 == 0xFFFFFFFF
